/analyze is not applied in CI

In order to pass `/analyze` to the build, the CI passed `-DCMAKE_CXX_FLAGS=/analyze:autolog-`. However, at https://github.com/microsoft/STL/blob/7f04137880e1c3df5125c1baf808f16f399dee2e/CMakeLists.txt#L43, we set `CMAKE_CXX_FLAGS` to `""`. This overrides the setting from the command line.

Thus, we aren't ever actually using `/analyze`. In fact, if we change the CMakeLists.txt to pass `/analyze`, we get warnings:

```
PS C:\..\nimazzuc\projects\stl> cmake --build out/test-analyze
[0/1] Re-running CMake...
-- Boost.Math: standalone mode ON
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/nimazzuc/projects/stl/out/test-analyze
[9/984] Building CXX object stl\CMakeFiles\msvcp_objects.dir\src\StlLCMapStringA.cpp.obj
FAILED: stl/CMakeFiles/msvcp_objects.dir/src/StlLCMapStringA.cpp.obj
C:\PROGRA~1\MICROS~1\2022\Preview\VC\Tools\MSVC\1433~1.315\bin\HOSTAR~1\ARM64\cl.exe  /nologo /TP -DCRTDLL2 -DNTDDI_VERSION=NTDDI_WIN10_CO -DSTRICT -DWIN32_LEAN_AND_MEAN -D_ALLOW_ITERATOR_DEBUG_LEVEL_MISMATCH -D_ARM64_ -D_CRTBLD -D_CRT_DECLARE_NONSTDC_NAMES=1 -D_CRT_STDIO_ARBITRARY_WIDE_SPECIFIERS -D_DLL -D_HAS_OLD_IOSTREAMS_MEMBERS=1 -D_STL_CONCRT_SUPPORT -D_STL_WIN32_WINNT=0x0A00 -D_VCRT_ALLOW_INTERNALS -D_VCRT_WIN32_WINNT=0x0A00 -D_WIN32_WINNT=0x0A00 -I"C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31517\crt\src\vcruntime" -IC:\Users\nimazzuc\projects\stl\stl\inc /WX /Gy /diagnostics:caret /W4 /w14265 /w15038 /fastfail /guard:cf /Z7 /Zp8 /std:c++latest /permissive- /Zc:threadSafeInit- /Zl /analyze:autolog- /O2 /Os /GL /EHsc /showIncludes /Fostl\CMakeFiles\msvcp_objects.dir\src\StlLCMapStringA.cpp.obj /Fdstl\CMakeFiles\msvcp_objects.dir\ /FS -c C:\Users\nimazzuc\projects\stl\stl\src\StlLCMapStringA.cpp
C:\Users\nimazzuc\projects\stl\stl\src\StlLCMapStringA.cpp(121,1): error C2220: the following warning is treated as an error    
}
^
c:\program files (x86)\windows kits\10\include\10.0.22000.0\um\winnls.h(3087) : warning C6553: The annotation for function 'LCMapStringEx' on _Param_(9) does not apply to a value type.
[11/984] Building CXX object stl\CMakeFiles\msvcp_objects.dir\src\StlLCMapStringW.cpp.obj
FAILED: stl/CMakeFiles/msvcp_objects.dir/src/StlLCMapStringW.cpp.obj
C:\PROGRA~1\MICROS~1\2022\Preview\VC\Tools\MSVC\1433~1.315\bin\HOSTAR~1\ARM64\cl.exe  /nologo /TP -DCRTDLL2 -DNTDDI_VERSION=NTDDI_WIN10_CO -DSTRICT -DWIN32_LEAN_AND_MEAN -D_ALLOW_ITERATOR_DEBUG_LEVEL_MISMATCH -D_ARM64_ -D_CRTBLD -D_CRT_DECLARE_NONSTDC_NAMES=1 -D_CRT_STDIO_ARBITRARY_WIDE_SPECIFIERS -D_DLL -D_HAS_OLD_IOSTREAMS_MEMBERS=1 -D_STL_CONCRT_SUPPORT -D_STL_WIN32_WINNT=0x0A00 -D_VCRT_ALLOW_INTERNALS -D_VCRT_WIN32_WINNT=0x0A00 -D_WIN32_WINNT=0x0A00 -I"C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.33.31517\crt\src\vcruntime" -IC:\Users\nimazzuc\projects\stl\stl\inc /WX /Gy /diagnostics:caret /W4 /w14265 /w15038 /fastfail /guard:cf /Z7 /Zp8 /std:c++latest /permissive- /Zc:threadSafeInit- /Zl /analyze:autolog- /O2 /Os /GL /EHsc /showIncludes /Fostl\CMakeFiles\msvcp_objects.dir\src\StlLCMapStringW.cpp.obj /Fdstl\CMakeFiles\msvcp_objects.dir\ /FS -c C:\Users\nimazzuc\projects\stl\stl\src\StlLCMapStringW.cpp
C:\Users\nimazzuc\projects\stl\stl\src\StlLCMapStringW.cpp(51,1): error C2220: the following warning is treated as an error     
}
^
c:\program files (x86)\windows kits\10\include\10.0.22000.0\um\winnls.h(3087) : warning C6553: The annotation for function 'LCMapStringEx' on _Param_(9) does not apply to a value type.
[14/984] Building CXX object stl\CMakeFiles\msvcp_objects.dir\src\instances.cpp.obj
ninja: build stopped: subcommand failed.
PS C:\..\nimazzuc\projects\stl>
```

Note that this is a bug in the Windows SDK - it applies `_In_opt_` to an `LPARAM` (i.e., an `intptr_t`)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -92,6 +92,7 @@ if(STL_USE_ANALYZE)
     # does not apply to a value type.
     # There's a bug in the declaration for LCMapStringEx - it applies _In_opt_ to an LPARAM.
     # LPARAM is a LONG_PTR (intptr_t), and it's invalid to apply _In_opt_ to a non-pointer.
+    # As of the Windows 11 SDK (10.0.22000.0), there are 5 total occurrences of warning C6553 affecting the STL's build.
     add_compile_options("$<$<COMPILE_LANGUAGE:CXX>:/analyze:autolog-;/wd6553>")
 endif()
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -40,6 +40,8 @@ endif()
 option(BUILD_TESTING "Enable testing" ON)
 set(VCLIBS_SUFFIX "_oss" CACHE STRING "suffix for built DLL names to avoid conflicts with distributed DLLs")
 
+option(STL_USE_ANALYZE "Pass the /analyze flag to MSVC" OFF)
+
 set(CMAKE_CXX_FLAGS "")
 set(CMAKE_CXX_FLAGS_DEBUG "")
 set(CMAKE_CXX_FLAGS_RELEASE "")
@@ -84,6 +86,19 @@ add_compile_options(/WX /Gy
     "$<$<COMPILE_LANGUAGE:CXX>:/diagnostics:caret;/W4;/w14265;/w15038;/fastfail;/guard:cf;/Z7;/Zp8;/std:c++latest;/permissive-;/Zc:threadSafeInit-;/Zl>"
     # note that /Zi generates debug info inside the object file, it's the same as /Z7 for msvc
     "$<$<COMPILE_LANGUAGE:ASM_MASM>:/Zi;/W3;/nologo>")
+
+if(STL_USE_ANALYZE)
+    # warning C6553: The annotation for function 'LCMapStringEx' on _Param_(9)
+    # does not apply to a value type.
+    # There's a bug in the declaration for LCMapStringEx - it applies _In_opt_ to an LPARAM.
+    # LPARAM is a LONG_PTR (intptr_t), and it's invalid to apply _In_opt_ to a non-pointer.
+
+    # warning C28182: Dereferencing NULL pointer. 'Temp_value_#1457' contains the same NULL value as
+    # 'new(1*40, nothrow)' did.
+    # `new` expressions with allocation functions that may return `null` are incorrectly treated as dereferencing
+    # the new'd pointer. This is a bug in C28182.
+    add_compile_options("$<$<COMPILE_LANGUAGE:CXX>:/analyze:autolog-;/wd6553;/wd28182>")
+endif()
 
 set(VCLIBS_DEBUG_OPTIONS "$<$<COMPILE_LANGUAGE:CXX>:/Od>")
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -11,6 +11,7 @@ check_cxx_source_compiles([=[
 static_assert(WDK_NTDDI_VERSION >= NTDDI_WIN10_CO, "Inspecting WDK_NTDDI_VERSION, the Windows SDK version.");
 int main() {}
 ]=] WINDOWS_SDK_VERSION_CHECK)
+
 if(NOT WINDOWS_SDK_VERSION_CHECK)
     message(FATAL_ERROR "The STL must be built with the Windows 11 SDK (10.0.22000.0) or later. Make sure it's available by selecting it in the Individual Components tab of the VS Installer.")
 endif()
@@ -18,6 +19,7 @@ endif()
 if(NOT DEFINED VCLIBS_TARGET_ARCHITECTURE)
     set(VCLIBS_TARGET_ARCHITECTURE "${CMAKE_CXX_COMPILER_ARCHITECTURE_ID}")
 endif()
+
 string(TOLOWER "${VCLIBS_TARGET_ARCHITECTURE}" VCLIBS_TARGET_ARCHITECTURE)
 
 if(VCLIBS_TARGET_ARCHITECTURE MATCHES "^x64$|^x86$")
@@ -26,12 +28,14 @@ endif()
 
 # add the tools subdirectory _before_ we change all the flags
 add_subdirectory(tools EXCLUDE_FROM_ALL)
+
 # these allow the targets to show up in the top-level
 # (as opposed to under the tools subdirectory)
 if(TARGET run-format)
     add_custom_target(format)
     add_dependencies(format run-format)
 endif()
+
 if(TARGET run-validate)
     add_custom_target(validate)
     add_dependencies(validate run-validate)
@@ -53,6 +57,7 @@ if("${VCLIBS_TARGET_ARCHITECTURE}" MATCHES "^x86$")
     set(VCLIBS_TARGET_ARCHITECTURE "x86")
     set(VCLIBS_I386_OR_AMD64 "i386")
     set(VCLIBS_X86_OR_X64 "x86")
+
     # Note that we set _WIN32_WINNT to a high level to make declarations available, but still engage downlevel
     # runtime dynamic linking by setting our own _STL_WIN32_WINNT back to Windows XP.
     add_compile_definitions(_X86_ _VCRT_WIN32_WINNT=0x0501 _STL_WIN32_WINNT=0x0501)
@@ -84,10 +89,12 @@ add_compile_definitions(
 
 add_compile_options(/WX /Gy
     "$<$<COMPILE_LANGUAGE:CXX>:/diagnostics:caret;/W4;/w14265;/w15038;/fastfail;/guard:cf;/Z7;/Zp8;/std:c++latest;/permissive-;/Zc:threadSafeInit-;/Zl>"
+
     # note that /Zi generates debug info inside the object file, it's the same as /Z7 for msvc
     "$<$<COMPILE_LANGUAGE:ASM_MASM>:/Zi;/W3;/nologo>")
 
 if(STL_USE_ANALYZE)
+    # TRANSITION OS-40109504: Windows SDK: incorrect SAL annotations on functions the STL uses
     # warning C6553: The annotation for function 'LCMapStringEx' on _Param_(9)
     # does not apply to a value type.
     # There's a bug in the declaration for LCMapStringEx - it applies _In_opt_ to an LPARAM.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -92,12 +92,7 @@ if(STL_USE_ANALYZE)
     # does not apply to a value type.
     # There's a bug in the declaration for LCMapStringEx - it applies _In_opt_ to an LPARAM.
     # LPARAM is a LONG_PTR (intptr_t), and it's invalid to apply _In_opt_ to a non-pointer.
-
-    # warning C28182: Dereferencing NULL pointer. 'Temp_value_#1457' contains the same NULL value as
-    # 'new(1*40, nothrow)' did.
-    # `new` expressions with allocation functions that may return `null` are incorrectly treated as dereferencing
-    # the new'd pointer. This is a bug in C28182.
-    add_compile_options("$<$<COMPILE_LANGUAGE:CXX>:/analyze:autolog-;/wd6553;/wd28182>")
+    add_compile_options("$<$<COMPILE_LANGUAGE:CXX>:/analyze:autolog-;/wd6553>")
 endif()
 
 set(VCLIBS_DEBUG_OPTIONS "$<$<COMPILE_LANGUAGE:CXX>:/Od>")

--- a/azure-devops/cmake-configure-build.yml
+++ b/azure-devops/cmake-configure-build.yml
@@ -33,7 +33,7 @@ steps:
     -DCMAKE_CXX_COMPILER=cl ^
     -DCMAKE_BUILD_TYPE=Release ^
     -DLIT_FLAGS=$(litFlags) ^
-    -DCMAKE_CXX_FLAGS=/analyze:autolog- ^
+    -DSTL_USE_ANALYZE=ON ^
     -S $(Build.SourcesDirectory) -B $(${{ parameters.buildOutputLocationVar }})
   displayName: 'Configure the STL'
   timeoutInMinutes: 2

--- a/stl/src/tzdb.cpp
+++ b/stl/src/tzdb.cpp
@@ -342,8 +342,11 @@ _NODISCARD __std_tzdb_time_zones_info* __stdcall __std_tzdb_get_time_zones() noe
     //    _Info == nullptr          --> bad_alloc
     //    _Info->_Err == _Win_error --> failed, call GetLastError()
     //    _Info->_Err == _Icu_error --> runtime_error interacting with ICU
+#pragma warning(push)
+#pragma warning(disable : 28182) // TRANSITION, VSO-1558277: Dereferencing NULL pointer.
     _STD unique_ptr<__std_tzdb_time_zones_info, _STD _Tzdb_deleter<__std_tzdb_time_zones_info>> _Info{
         new (_STD nothrow) __std_tzdb_time_zones_info{}};
+#pragma warning(pop)
     if (_Info == nullptr) {
         return nullptr;
     }
@@ -440,8 +443,11 @@ _NODISCARD __std_tzdb_current_zone_info* __stdcall __std_tzdb_get_current_zone()
     //    _Info == nullptr          --> bad_alloc
     //    _Info->_Err == _Win_error --> failed, call GetLastError()
     //    _Info->_Err == _Icu_error --> runtime_error interacting with ICU
+#pragma warning(push)
+#pragma warning(disable : 28182) // TRANSITION, VSO-1558277: Dereferencing NULL pointer.
     _STD unique_ptr<__std_tzdb_current_zone_info, _STD _Tzdb_deleter<__std_tzdb_current_zone_info>> _Info{
         new (_STD nothrow) __std_tzdb_current_zone_info{}};
+#pragma warning(pop)
     if (_Info == nullptr) {
         return nullptr;
     }
@@ -478,8 +484,11 @@ _NODISCARD __std_tzdb_sys_info* __stdcall __std_tzdb_get_sys_info(
     //    _Info == nullptr          --> bad_alloc
     //    _Info->_Err == _Win_error --> failed, call GetLastError()
     //    _Info->_Err == _Icu_error --> runtime_error interacting with ICU
+#pragma warning(push)
+#pragma warning(disable : 28182) // TRANSITION, VSO-1558277: Dereferencing NULL pointer.
     _STD unique_ptr<__std_tzdb_sys_info, _STD _Tzdb_deleter<__std_tzdb_sys_info>> _Info{
         new (_STD nothrow) __std_tzdb_sys_info{}};
+#pragma warning(pop)
     if (_Info == nullptr) {
         return nullptr;
     }

--- a/stl/src/winapisupp.cpp
+++ b/stl/src/winapisupp.cpp
@@ -193,7 +193,7 @@ extern "C" _CRTIMP2 PTP_TIMER __cdecl __crtCreateThreadpoolTimer(
 
 // TRANSITION, ABI: preserved for binary compatibility
 extern "C" _CRTIMP2 VOID __cdecl __crtSetThreadpoolTimer(_Inout_ PTP_TIMER const pti,
-    _In_opt_ PFILETIME const pftDueTime, _In_ DWORD const msPeriod, _In_opt_ DWORD const msWindowLength) {
+    _In_opt_ PFILETIME const pftDueTime, _In_ DWORD const msPeriod, _In_ DWORD const msWindowLength) {
     SetThreadpoolTimer(pti, pftDueTime, msPeriod, msWindowLength);
 }
 

